From 0dd09ae7f460835058d6e13ed75370e5007614d2 Mon Sep 17 00:00:00 2001
From: Michal Sekletar <msekleta@redhat.com>
Date: Wed, 7 Oct 2026 16:39:17 +0200
Subject: [PATCH] avahi-daemon/dbus: process only NameOwnerChanged signals
 originated from the bus broker

Let's verify sender of the signal before acting on the message. This
avoids unnecessary processing and potential cleanup of wrong avahi
client based on bogus signal generated by other/malicious D-Bus clients.
---
 avahi-daemon/dbus-protocol.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/avahi-daemon/dbus-protocol.c b/avahi-daemon/dbus-protocol.c
index 2765cf3..e0a4e03 100644
--- a/avahi-daemon/dbus-protocol.c
+++ b/avahi-daemon/dbus-protocol.c
@@ -220,6 +220,11 @@ static DBusHandlerResult msg_signal_filter_impl(AVAHI_GCC_UNUSED DBusConnection
 
     } else if (dbus_message_is_signal(m, DBUS_INTERFACE_DBUS, "NameOwnerChanged")) {
         char *name, *old, *new;
+        const char *sender;
+
+        sender = dbus_message_get_sender(m);
+        if (!sender || strcmp(sender, DBUS_SERVICE_DBUS) != 0)
+            return DBUS_HANDLER_RESULT_HANDLED;
 
         if (!dbus_message_get_args(m, &error, DBUS_TYPE_STRING, &name, DBUS_TYPE_STRING, &old, DBUS_TYPE_STRING, &new, DBUS_TYPE_INVALID)) {
             return dbus_parsing_error("Error parsing NameOwnerChanged message", &error);
